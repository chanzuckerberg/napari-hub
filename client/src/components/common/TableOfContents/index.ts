export * from './TableOfContents';
export * from './TableOfContents.constants';
export * from './TableOfContents.types';
