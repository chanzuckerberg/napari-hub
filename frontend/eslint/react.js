module.exports = {
  extends: ['airbnb/hooks'],

  plugins: ['napari-hub'],

  rules: {
    'napari-hub/no-root-mui-import': 'error',

    /*
      Prop types aren't necessary since we have TypeScript interfaces for
      prop types.
    */
    'react/prop-types': 'off',

    /*
      React automatically adds the import as part of the new JSX transform:
      https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html#nextjs
     */
    'react/react-in-jsx-scope': 'off',

    /*
      Prop spreading is very useful as a simple way for passing multiple
      props to a component. However, we should still be careful about passing
      unnecessary props.
    */
    'react/jsx-props-no-spreading': 'off',

    /*
      This rule isn't really needed anymore since we can use ES6 object defaults
      for props. There are also plans to deprecate the use of `defaultProps`:
      https://github.com/reactjs/rfcs/pull/107
    */
    'react/require-default-props': 'off',
  },
};
