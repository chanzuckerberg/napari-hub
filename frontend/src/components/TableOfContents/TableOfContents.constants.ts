/**
 * Header tag to use for TOC.
 */
export const TOC_HEADER_TAG = 'h2';
