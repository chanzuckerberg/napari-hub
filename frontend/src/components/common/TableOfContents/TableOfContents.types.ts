export interface TOCHeader {
  id: string;
  text: string;
}
