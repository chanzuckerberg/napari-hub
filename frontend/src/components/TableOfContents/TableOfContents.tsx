import clsx from 'clsx';

import { useActiveHeader } from './TableOfContents.hooks';
import { TOCHeader } from './TableOfContents.types';

interface Props {
  /**
   * className: additional classes to apply to this component
   */
  className?: string;

  /**
   * onClick: callback for when headings are clicked.
   */
  onClick?(heading: string): void;

  /**
   * headers: header ids and titles to link to
   */
  headers: TOCHeader[];

  /**
   * free: whether the component should move with the page or be fixed in place
   */
  free?: boolean;
}

/**
 * Duration to wait for before re-enabling the TOC event handlers.
 */
const ENABLE_EVENT_HANDLERS_TIMEOUT_MS = 100;

/**
 * Component for rendering TOC from the given headers. Highlighting will
 * only work if the headers match those present on the page.
 */
export function TableOfContents({ className, onClick, headers, free }: Props) {
  const {
    activeHeader,
    setActiveHeader,
    enableEventHandlers,
    disableEventHandlers,
  } = useActiveHeader(headers);

  return (
    <ul
      className={clsx(
        className,
        'flex flex-col',
        'border-l border-black',
        !free && 'sticky top-12',
      )}
    >
      {headers.map((header) => {
        const isActive = header.id === activeHeader;

        return (
          <li
            className={clsx(
              // Layout
              'flex',
              // 'flex items-center',

              // Box model
              'pl-sds-xl border-l-4',

              // Ensure height is at least 25px, but also allow scaling for
              // multiline items.
              'min-h-6 h-auto',

              // Apply top/bottom margins except for first/last items
              'my-sds-s first:mt-0 last:mb-0',

              // Smooth transition for border color
              'transition-colors',

              'hover:border-napari-primary',
              !isActive && 'border-transparent',
              isActive && 'border-black',
            )}
            key={header.id}
            data-active={isActive}
            data-testid="tocItem"
          >
            {/*
              Use normal link component instead of Next.js Link because we're
              not loading another page.
            */}
            <a
              className={clsx(isActive && 'font-bold')}
              href={`#${header.id}`}
              onClick={(event) => {
                event.preventDefault();

                // The event handlers are disabled here because we want to set
                // the active header AND scroll to the header. If the handlers
                // aren't disabled, then the handlers and the below
                // `setActiveHeader()` will have a race condition.
                disableEventHandlers();

                // Set the hash to the header ID so that the page scrolls to it.
                window.location.hash = header.id;
                setActiveHeader(header.id);
                onClick?.(header.text);

                // Wrap in timeout so that the browser has time to scroll the
                // header. If we don't wrap it in a timeout, then setting the
                // hash will fire a scroll event and overwrite the active
                // header.
                setTimeout(
                  () => enableEventHandlers(),
                  ENABLE_EVENT_HANDLERS_TIMEOUT_MS,
                );
              }}
            >
              {header.text}
            </a>
          </li>
        );
      })}
    </ul>
  );
}
