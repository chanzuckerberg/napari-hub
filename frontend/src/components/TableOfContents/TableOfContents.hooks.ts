import { throttle } from 'lodash';
import { useCallback, useEffect, useState } from 'react';

import { TOCHeader } from './TableOfContents.types';

/**
 * The designs requires a 35px margin between the heading and the top of the
 * viewport when scrolling to a heading, so the next heading should be
 * highlighted when hitting that offset.
 */
const TOP_OFFSET_PX = 35;

/**
 * Duration of throttling for  `findActiveHeader()` function. This is so that
 * the `findActiveHeader()` function is called no more than 100ms so that we can
 * reduce expensive calls to `getBoundingClientRect()`.
 */
const FIND_ACTIVE_HEADER_THROTTLE_MS = 100;

/**
 * Duration to wait for before setting the active header on initial load.
 */
const SET_INITIAL_HEADER_TIMEOUT_MS = 200;

/**
 * Utility function for finding the first header in the viewport by looking for
 * the first positive `top` value.
 *
 * TODO Look into using IntersectionObserver API: https://mzl.la/3lYJvpt
 *
 * @param headers The markdown headers.
 * @returns Data for the first header in the viewport.
 */
function firstHeaderInViewport(headers: TOCHeader[]) {
  // Find first header that is in viewport.
  let firstHeaderTop = 0;
  let firstHeaderIndex = -1;

  for (let i = 0; i < headers.length; i += 1) {
    const { id } = headers[i];
    const node = document.getElementById(id);
    const top = node?.getBoundingClientRect()?.top ?? 0;

    if (top >= 0) {
      firstHeaderTop = top;
      firstHeaderIndex = i;
      break;
    }
  }

  const firstHeader = headers[firstHeaderIndex] as TOCHeader | undefined;

  return {
    firstHeader,
    firstHeaderTop,
    firstHeaderIndex,
  };
}

/**
 * Hook for getting the selected heading anchor from an array of markdown
 * headers.  The headers should be rendered on the DOM via the <Markdown />
 * component for this to work.
 *
 * This code is inspired by the TOC for Docusaurus: https://git.io/JObbd, with
 * small improvements.
 *
 * @param headers Markdown headers from `getHeadersFromMarkdown()`
 * @returns Active header ID
 */
export function useActiveHeader(headers: TOCHeader[]) {
  const [activeHeader, setActiveHeader] = useState(headers[0]?.id ?? '');

  // Function that finds the active header based on the current viewport `scrollY`.
  // eslint-disable-next-line react-hooks/exhaustive-deps
  const findActiveHeader = useCallback(
    throttle(() => {
      const { firstHeader, firstHeaderIndex, firstHeaderTop } =
        firstHeaderInViewport(headers);

      if (!firstHeader) {
        return;
      }

      if (Math.floor(firstHeaderTop) <= TOP_OFFSET_PX) {
        setActiveHeader(firstHeader.id);
      } else {
        // If the first header in the viewport is greater than the offset, then the
        // user is still in the previous section.
        const previousHeader = headers[firstHeaderIndex - 1] ?? firstHeader;
        setActiveHeader(previousHeader.id);
      }
    }, FIND_ACTIVE_HEADER_THROTTLE_MS),
    [headers],
  );

  const enableEventHandlers = useCallback(() => {
    document.addEventListener('scroll', findActiveHeader);
    document.addEventListener('resize', findActiveHeader);
  }, [findActiveHeader]);

  const disableEventHandlers = useCallback(() => {
    document.removeEventListener('scroll', findActiveHeader);
    document.removeEventListener('resize', findActiveHeader);
  }, [findActiveHeader]);

  // Effect for determining the active header on initial load. For some reason,
  // this effect doesn't work on Safari because of some weird hash link issue.
  // When opening a URL like https://napari-hub.org/plugins/example#license,
  // Safari will sometimes open the page several pixels above the `License`
  // header.
  //
  // The same issue can be seen on GitHub when opening a hash link in Safari:
  // https://docs.github.com/en/actions/creating-actions/about-actions#creating-a-readme-file-for-your-action
  // TODO Look into fix for Safari.
  useEffect(() => {
    const initialHeader = window.location.hash.replace(/^#/, '');

    // If the user is opening a link with a header specified in the hash, then
    // we need to figure out the correct active header to render.
    if (headers.find((header) => header.id === initialHeader)) {
      // Wrap in timeout so that the DOM has time to settle after initial load.
      setTimeout(() => {
        // If the header is at the bottom of the page, then set the active
        // header to whatever was specified in the URL hash.
        if (window.innerHeight + window.scrollY >= document.body.scrollHeight) {
          setActiveHeader(initialHeader);
        } else {
          // If the page is loaded anywhere else, then just find the next active
          // header in the viewport.
          findActiveHeader();
        }

        enableEventHandlers();
      }, SET_INITIAL_HEADER_TIMEOUT_MS);
    } else {
      // If there is no header specified in the hash, then just find the next
      // active header in the viewport.
      findActiveHeader();
      enableEventHandlers();
    }

    return disableEventHandlers;
  }, [disableEventHandlers, enableEventHandlers, findActiveHeader, headers]);

  return {
    activeHeader,
    setActiveHeader,
    enableEventHandlers,
    disableEventHandlers,
  };
}
