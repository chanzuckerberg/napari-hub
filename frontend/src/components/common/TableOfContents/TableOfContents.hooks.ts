import { useEffect, useState } from 'react';

import { TOCHeader } from './TableOfContents.types';

/**
 * The designs requires a 35px margin between the heading and the top of the
 * viewport when scrolling to a heading, so the next heading should be
 * highlighted when hitting that offset.
 */
const TOP_OFFSET = 35;

/**
 * Hook for getting the selected heading anchor from an array of markdown
 * headers.  The headers should be rendered on the DOM via the <Markdown />
 * component for this to work.
 *
 * This code is inspired by the TOC for Docusaurus: https://git.io/JObbd, with
 * small improvements.
 *
 * @param headers Markdown headers from `getHeadersFromMarkdown()`
 * @returns Active header ID
 */
export function useActiveHeader(headers: TOCHeader[]): string {
  const [active, setActive] = useState(headers[0]?.id ?? '');

  useEffect(() => {
    function findActiveHeader() {
      // Get headers as DOM nodes.
      const headerTags = headers.map((header) =>
        document.getElementById(header.id),
      );

      // Find first header that is in viewport.
      const firstHeaderIndex = headerTags.findIndex((header) => {
        const top = header?.getBoundingClientRect()?.top ?? 0;
        return top >= 0;
      });
      const firstHeader = headerTags[firstHeaderIndex];

      if (firstHeader) {
        const { top } = firstHeader.getBoundingClientRect();

        // If user reaches the bottom, set the last header as selected.
        if (
          window.innerHeight + window.pageYOffset >=
          document.body.scrollHeight
        ) {
          setActive(headers[headers.length - 1].id);
        } else if (Math.floor(top) <= TOP_OFFSET) {
          setActive(firstHeader.id);
        } else {
          /*
            If the first header in viewport is greater than the offset, then
            the user is still in the previous section.
          */
          const previousHeader = headers[firstHeaderIndex - 1] ?? firstHeader;
          setActive(previousHeader.id);
        }
      }
    }

    document.addEventListener('scroll', findActiveHeader);
    document.addEventListener('resize', findActiveHeader);

    // Find active header on initial render
    findActiveHeader();

    // Remove event listeners on cleanup.
    return () => {
      document.removeEventListener('scroll', findActiveHeader);
      document.removeEventListener('resize', findActiveHeader);
    };
  }, [headers]);

  return active;
}
