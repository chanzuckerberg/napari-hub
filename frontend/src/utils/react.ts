import { ReactElement, ReactNode } from 'react';

/**
 * Checks if a React node is an element. An element is an instantiated
 * component.
 *
 * @param node The React node
 * @returns True if node is an element
 */
export function isReactElement(node: ReactNode): node is ReactElement {
  const element = node as ReactElement | null | undefined;
  return !!(element?.type && element?.props);
}
