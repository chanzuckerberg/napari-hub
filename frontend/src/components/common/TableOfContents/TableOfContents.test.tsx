import { render } from '@testing-library/react';

import { TableOfContents } from './TableOfContents';

describe('Table of Contents', () => {
  it('should match snapshot', () => {
    const { asFragment } = render(
      <TableOfContents
        headers={[
          { id: 'one', text: 'One' },
          { id: 'two', text: 'Two' },
        ]}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  // TODO: test highlighting
});
