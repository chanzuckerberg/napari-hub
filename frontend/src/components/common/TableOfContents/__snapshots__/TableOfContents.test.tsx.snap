// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table of Contents should match snapshot 1`] = `
<DocumentFragment>
  <ul
    class="flex flex-col border-l border-black sticky top-12"
  >
    <li
      class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-black"
      data-active="true"
      data-testid="tocItem"
    >
      <a
        class="font-bold"
        href="#one"
      >
        One
      </a>
    </li>
    <li
      class="flex pl-6 border-l-4 min-h-6 h-auto my-2 first:mt-0 last:mb-0 transition-colors hover:border-napari-primary border-transparent"
      data-active="false"
      data-testid="tocItem"
    >
      <a
        class=""
        href="#two"
      >
        Two
      </a>
    </li>
  </ul>
</DocumentFragment>
`;
